contracts/peripheral/YieldDirectorV2.sol

```diff
@@ -0,0 +1,441 @@
+        @param id_ Deposit ID to remove gOHM debt from
+        @param amount_ Amount of gOHM debt to remove and return to donor
+     */
+    function withdrawPrincipal(uint256 id_, uint256 amount_) external {
+        require(isValidWithdrawal(amount_), "Not a valid withdrawal.");
+        DepositInfo storage currDeposit = depositInfo[id_];
+        require(currDeposit.depositor == msg.sender, "Can't withdraw principal that's not yours");
+
+        if (amount_ >= IgOHM(gOHM).balanceTo(currDeposit.principalAmount)) {
+            _withdrawPrincipal(id_, amount_);
+            currDeposit.principalAmount = 0;
+            emit Donated(
+                msg.sender,
+                recipientLookup[id_],
+                _getOutstandingYield(currDeposit.principalAmount, currDeposit.agnosticAmount)
+            );
+        } else {
+            _withdrawPrincipal(id_, amount_);
+        }
+
+        IERC20(gOHM).safeTransfer(msg.sender, amount_);
+        emit Withdrawn(msg.sender, recipientLookup[id_], amount_);
+    }
+
+    /**
+        @notice Withdraw donor's gOHM from vault, and return it as sOHM
+        @param id_ Deposit ID to remove gOHM debt from
+        @param amount_ Amount of gOHM debt to remove and return to donor as sOHM
+     */
+    function withdrawPrincipalAsSohm(uint256 id_, uint256 amount_) external {
+        require(isValidWithdrawal(amount_), "Not a valid withdrawal.");
+        DepositInfo storage currDeposit = depositInfo[id_];
+        require(currDeposit.depositor == msg.sender, "Can't withdraw principal that's not yours");
+
+        if (amount_ >= IgOHM(gOHM).balanceTo(currDeposit.principalAmount)) {
+            _withdrawPrincipal(id_, amount_);
+            currDeposit.principalAmount = 0;
+            emit Donated(
+                msg.sender,
+                recipientLookup[id_],
+                _getOutstandingYield(currDeposit.principalAmount, currDeposit.agnosticAmount)
+            );
+        } else {
+            _withdrawPrincipal(id_, amount_);
+        }
+
+        IERC20(sOHM).approve(address(staking), amount_);
+        staking.unwrap(msg.sender, amount_);
+
+        emit Withdrawn(msg.sender, recipientLookup[id_], amount_);
+    }
+
+    /**
+        @notice Withdraw from all donor positions
+     */
+    function withdrawAll() external {
+        require(!withdrawDisabled, "Withdraws currently disabled");
+
+        uint256[] storage depositIds = depositorIds[msg.sender];
+
+        uint256 depositsLength = depositIds.length;
+        require(depositsLength != 0, "User not donating to anything");
+
+        uint256 principalTotal = 0;
+
+        for (uint256 index = 0; index < depositsLength; index++) {
+            DepositInfo storage currDeposit = depositInfo[depositIds[index]];
+            principalTotal += currDeposit.principalAmount;
+            _withdrawPrincipal(currDeposit.id, IgOHM(gOHM).balanceTo(currDeposit.principalAmount));
```

Comment:
would it be helpful if base contract had a function _withdrawAllprincipal(uint256 id). Thats withdraws all the principal and returns the amount.

Seems reasonably generic to add to base contract to me. What u think.

Reply:
Yeah think that would be useful and generic enough to add

Reply:
I think im going to make a pr with yieldsplitter as only file change that way we can both rebase on it without having to coodinate with each other on all the changes,